src_features/provideDomainName/cmd_provide_domain_name.c

```diff
@@ -0,0 +1,567 @@
+            case TLV_TAG:
+                data.tag = payload->ptr[offset];
+                offset += sizeof(data.tag);
+                step = TLV_LENGTH;
+                break;
+
+            case TLV_LENGTH:
+                data.length = payload->ptr[offset];
+                offset += sizeof(data.length);
+                step = TLV_VALUE;
+                break;
+
+            case TLV_VALUE:
+                if (offset >= payload->size) {
+                    apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                    return false;
+                }
+                data.value = &payload->ptr[offset];
+                // check if a handler exists for this tag
+                for (int idx = 0; idx < (int) ARRAY_SIZE(handlers); ++idx) {
+                    if (handlers[idx].tag == data.tag) {
+                        handlers[idx].found += 1;
+                        t_tlv_handler fptr = PIC(handlers[idx].func);
+                        if (!(*fptr)(&data, domain_name_info, sig_ctx)) {
+                            PRINTF("Error while handling tag 0x%x\n", handlers[idx].tag);
+                            apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                            return false;
+                        }
+                        break;
+                    }
+                }
+                if (data.tag != SIGNATURE) {  // the signature wasn't computed on itself
+                    feed_tlv_hash(&data, sig_ctx);
+                }
+                offset += data.length;
+                step = TLV_TAG;
+                break;
+
+            default:
+                apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                return false;
+        }
+    }
+    // prevent missing or duplicated tags
+    for (int idx = 0; idx < (int) ARRAY_SIZE(handlers); ++idx) {
+        if (handlers[idx].found != 1) {
+            PRINTF("Found %u occurence(s) of tag 0x%x in TLV!\n",
+                   handlers[idx].found,
+                   handlers[idx].tag);
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Allocate and assign TLV payload
+ *
+ * @param[in] payload payload structure
+ * @param[in] size size of the payload
+ * @return whether it was successful
+ */
+static bool set_payload(s_tlv_payload *payload, uint16_t size) {
+    if ((payload->ptr = mem_alloc(size)) == NULL) {
+        apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+        return false;
+    }
+    payload->expected_size = size;
+    return true;
+}
+
+/**
+ * Deallocate and unassign TLV payload
+ *
+ * @param[in] payload payload structure
+ */
+static void unset_payload(s_tlv_payload *payload) {
+    mem_dealloc(payload->expected_size);
+    memset(payload, 0, sizeof(*payload));
+}
+
+static bool handle_first_chunk(const uint8_t **data, uint8_t *length, s_tlv_payload *payload) {
+    // check if no payload is already in memory
+    if (payload->ptr != NULL) {
+        unset_payload(payload);
```

Comment:
Why on a faulty APDU reception we cancel the flow ?

Reply:
If the host starts sending faulty APDUs, it's either trying to do something sketchy or is lost.
And for example, if the host crashed in the middle of a PROVIDE_DOMAIN_NAME, there would be no way for it to restart one if the app kept refusing and just waited for the proper follow up to the last one.